File: desktop/frontend/src/lib/starters.ts
// The four cards on an empty chat, for the rooms that are part of the product.
//
// One list of starters for the whole app was a card set written for the
// workshop shown in every other room: "รีวิวโค้ดและแนะนำการแก้ไข" on ผู้ช่วย,
// which has no code tools, and on a chat with the document agent, who would not
// know what to do with it. The empty state is the first sentence a room says,
// so it has to be that room's sentence.
//
// **The agents' own cards are NOT here.** A worker keeps its opening in its own
// folder, beside its AGENT.md (internal/subagent/starters.go), and this file
// never learns their names. That line is the whole design: an agent is a folder
// the user can add, so a card set that lived here could only ever be given to
// the workers we shipped — hiring a bookkeeper would mean editing the product
// to let it say hello (owner's call, 2026-08-10). What is left here is the
// rooms the owner draws: the two desks, a project, and the four cards any
// colleague opens with when its folder holds no opening of its own.
//
// The rule these sets are written under is the same one the desks are under: a
// card may only ask for work the session on the other side can actually do
// (docs/DECISIONS.md — "โต๊ะที่ไม่มีเครื่องมือ ต้องไม่ถูกสั่งให้ใช้มัน"). A
// starter is a pre-filled request, so an impossible one costs the user a turn
// and reads as the assistant's failure rather than the card's.
//
// **And "can do" is not the bar — "is good at" is** (owner's call, 2026-08-15).
// A trip-planning card shipped here briefly and was cut, not because the tools
// were missing on paper — it is web search and a written file, both present —
// but because a real trip plan turns on distances, opening hours and a map, and
// this product has no sense of place at all. Every tool the card touched
// existed; the thing it promised was still something we are bad at. So the test
// before a card goes in the pool is not "does a tool exist for each step", it
// is "would we be pleased to be judged on this" — the empty screen is where a
// user forms their opinion of what this app is for, and a card is a promise
// made before any work is done.
//
// The card fills the composer rather than sending — the prompts ending in ": "
// are deliberately unfinished, because the half a sentence only the user can
// write is exactly the half worth asking for.
//
// **A card is two pieces of writing with two different jobs** (owner's call,
// 2026-08-15). The title sells the outcome and is read in a second, so it stays
// short. The prompt is not a restatement of the title — it is the actual
// instruction the model receives, nobody has to read it, and every sentence
// added to it is quality in the work that comes back. So ผู้ช่วย's prompts are
// written the way a prompt engineer writes: they name the sources to use, the
// rule against inventing a number, the order of work, the exact artifact
// wanted, and what to say about where each finding came from. Writing them as
// one-line echoes of the title was the mistake this note exists to prevent.
//
// **ผู้ช่วย's cards name a real job, and are finished sentences** (owner's call,
// 2026-08-15). The four they replace named an ability and stopped — "อ่านและ
// สรุปเอกสารหรือสื่อ", "ค้นข้อมูลจากเว็บแล้วสรุปให้" — which is the sentence a
// chat tab in a browser also says, so the room that does the work on your
// machine was opening with the one claim that proves nothing. What replaced
// them was written under one rule: **a card ends in something you can open.**
// That is the line the owner drew when he kept two of a proposed six and cut
// the rest; the survivors both handed back a file, and every one he cut handed
// back a summary. It is also what this class of product actually says on its
// front page — Manus opens with "Create slides / Build website / Create games"
// and shows finished jobs with proper nouns ("Trip to Japan in april"), never
// "summarize a document".
//
// Naming a real subject rather than a shape ("โมเดล AI ตัวไหนคุ้มที่สุด", not
// "ค้นข้อมูลเรื่องหนึ่ง") costs the user a delete when they wanted a different
// subject, and buys the thing an empty screen is for: they see the whole job at
// once instead of decoding what this app is willing to be asked. The owner
// weighed that trade and took the picture.

import type { IconName } from './icons'
import type { TKey } from './i18n.svelte'

export interface Starter {
  icon: IconName
  titleKey: TKey
  promptKey: TKey
}

export interface StarterSet {
  /** The question above the cards. It belongs to the set: "จะให้เราสร้างอะไรดี"
   *  is the workshop asking, and ผู้ช่วย is not mainly asked to build. */
  headlineKey: TKey
  /** Every card the room could open with. NOT what is drawn — `dealStarters`
   *  hands the window four of these. A set may hold as many as it has. */
  starters: Starter[]
}

/** Keyed by context id: a desk name, or 'project'. `chair` is what any office
 *  agent opens with when its own folder holds no STARTERS.md. */
const SETS: Record<string, StarterSet> = {
  assistant: {
    headlineKey: 'start.assistant.headline',
    starters: [
      { icon: 'search', titleKey: 'start.assistant.chartTitle', promptKey: 'start.assistant.chartPrompt' },
      { icon: 'monitor', titleKey: 'start.assistant.healthTitle', promptKey: 'start.assistant.healthPrompt' },
      { icon: 'layoutList', titleKey: 'start.assistant.moneyTitle', promptKey: 'start.assistant.moneyPrompt' },
      { icon: 'terminal', titleKey: 'start.assistant.appTitle', promptKey: 'start.assistant.appPrompt' },
      { icon: 'globe', titleKey: 'start.assistant.scrapeTitle', promptKey: 'start.assistant.scrapePrompt' },
      { icon: 'brain', titleKey: 'start.assistant.learnTitle', promptKey: 'start.assistant.learnPrompt' },
      { icon: 'chartColumn', titleKey: 'start.assistant.bizTitle', promptKey: 'start.assistant.bizPrompt' },
      { icon: 'timer', titleKey: 'start.assistant.runwayTitle', promptKey: 'start.assistant.runwayPrompt' },
      { icon: 'compass', titleKey: 'start.assistant.marketTitle', promptKey: 'start.assistant.marketPrompt' },
      { icon: 'package', titleKey: 'start.assistant.deckTitle', promptKey: 'start.assistant.deckPrompt' },
    ],
  },

  coding: {
    headlineKey: 'start.coding.headline',
    starters: [
      { icon: 'bandage', titleKey: 'start.coding.fixTitle', promptKey: 'start.coding.fixPrompt' },
      { icon: 'wrench', titleKey: 'start.coding.buildTitle', promptKey: 'start.coding.buildPrompt' },
      { icon: 'compass', titleKey: 'start.coding.exploreTitle', promptKey: 'start.coding.explorePrompt' },
      { icon: 'search', titleKey: 'start.coding.reviewTitle', promptKey: 'start.coding.reviewPrompt' },
      { icon: 'package', titleKey: 'start.coding.upgradeTitle', promptKey: 'start.coding.upgradePrompt' },
      { icon: 'shield', titleKey: 'start.coding.testTitle', promptKey: 'start.coding.testPrompt' },
      { icon: 'timer', titleKey: 'start.coding.perfTitle', promptKey: 'start.coding.perfPrompt' },
      { icon: 'puzzle', titleKey: 'start.coding.refactorTitle', promptKey: 'start.coding.refactorPrompt' },
    ],
  },

  // A chat held inside a โปรเจกต์. What makes it different from a plain
  // assistant chat is exactly one thing — the context folder rides into every
  // session in it — so all four cards are about that folder. Nothing here may
  // refer to the project's *other* chats: a session is handed the files, not
  // the earlier conversations, and a card promising a recap would be asking for
  // something the session cannot see.
  project: {
    headlineKey: 'start.project.headline',
    starters: [
      { icon: 'folder', titleKey: 'start.project.filesTitle', promptKey: 'start.project.filesPrompt' },
      { icon: 'search', titleKey: 'start.project.askTitle', promptKey: 'start.project.askPrompt' },
      { icon: 'wrench', titleKey: 'start.project.continueTitle', promptKey: 'start.project.continuePrompt' },
      { icon: 'layoutList', titleKey: 'start.project.planTitle', promptKey: 'start.project.planPrompt' },
      { icon: 'alertTriangle', titleKey: 'start.project.conflictTitle', promptKey: 'start.project.conflictPrompt' },
      { icon: 'fileText', titleKey: 'start.project.briefTitle', promptKey: 'start.project.briefPrompt' },
    ],
  },

  // Any office agent whose folder holds no STARTERS.md — the shipped five all
  // carry one, so in practice this is a worker the user hired, or one whose file
  // is empty. The four cards name the four ways a conversation with a colleague
  // starts, which stays true of a specialist whose craft this file has never
  // heard of. That is the only thing the window is allowed to assume about an
  // agent it did not ship.
  chair: {
    headlineKey: 'start.chair.headline',
    starters: [
      { icon: 'messageSquare', titleKey: 'start.chair.whatTitle', promptKey: 'start.chair.whatPrompt' },
      { icon: 'brain', titleKey: 'start.chair.consultTitle', promptKey: 'start.chair.consultPrompt' },
      { icon: 'folderOpen', titleKey: 'start.chair.lookTitle', promptKey: 'start.chair.lookPrompt' },
      { icon: 'zap', titleKey: 'start.chair.doTitle', promptKey: 'start.chair.doPrompt' },
      { icon: 'search', titleKey: 'start.chair.checkTitle', promptKey: 'start.chair.checkPrompt' },
      { icon: 'package', titleKey: 'start.chair.prepTitle', promptKey: 'start.chair.prepPrompt' },
    ],
  },
}

/** Which room the open session is in, most specific first.
 *
 * The order is the session's own order of specificity, not the nav's: a chair
 * session is with a named agent and that beats everything, a project chat is an
 * assistant chat filed somewhere so the folder wins over the desk, and the desk
 * answers the rest. The '' desk — sessions from before desks existed, and any
 * moment before the engine has answered — lands on ผู้ช่วย, which is the desk
 * those sessions behave like.
 *
 * For a chair this is the FLOOR, not the answer: the agent's own folder is
 * asked first (ChairStarters), and what comes back replaces this whenever the
 * worker had something to say. */
export function startersFor(ctx: { desk: string; chair: string; space: string }): StarterSet {
  if (ctx.chair) return SETS.chair
  if (ctx.space) return SETS.project
  return SETS[ctx.desk] ?? SETS.assistant
}

/** How many cards the grid draws, whatever the room is holding. */
export const STARTER_SLOTS = 4

/** Deal `count` cards out of a pool, dealing every card before repeating any.
 *
 * The empty state shows four cards and a room may hold far more — the point of
 * a pool is that a set can keep growing without the screen growing with it
 * (owner's call, 2026-08-15). Four was drawn from a fixed list before, which
 * meant a fifth card could only be added by taking one away.
 *
 * **A shuffled deck, not a dice roll.** Rolling four at random can hide a card
 * for a dozen openings and show the same hand twice in a row, which is exactly
 * backwards: the reason to hold a pool is that people find things in it. So a
 * bag is kept per room, dealt from, and re-shuffled only once it runs out —
 * every card is seen before any card comes back.
 *
 * The bag lives here rather than on disk. It is a nicety about what you saw
 * this afternoon, not a fact about the product, and losing it on restart costs
 * exactly one reshuffle. A room whose pool is no bigger than a hand skips all
 * of this and hands the pool straight back.
 */
// The bag holds NAMES, not cards. The pool is rebuilt from scratch every time
// the window re-derives it — new objects, same words — so a bag of cards would
// be a bag of stale references that match nothing in the pool it is supposed to
// be dealing from, and every "have I dealt this already" check would answer no.
// (It did. The grid drew the same card twice.) A name survives the rebuild, and
// dropping names the pool no longer has is also how a pool that CHANGED — the
// language switched, the agent's file arrived — heals itself instead of dealing
// cards that no longer exist.
const bags = new Map<string, string[]>()

export function dealStarters<T>(
  key: string,
  pool: readonly T[],
  nameOf: (card: T) => string,
  count = STARTER_SLOTS,
): T[] {
  if (pool.length <= count) return [...pool]

  const byName = new Map(pool.map((c) => [nameOf(c), c]))
  const hand: T[] = []
  const dealt = new Set<string>()
  let bag = (bags.get(key) ?? []).filter((n) => byName.has(n))

  while (hand.length < count) {
    // Re-shuffle excluding what is already in this hand, so a bag running dry
    // mid-deal cannot put the same card on screen twice.
    if (bag.length === 0) bag = shuffle([...byName.keys()].filter((n) => !dealt.has(n)))
    const name = bag.pop() as string
    if (dealt.has(name)) continue
    dealt.add(name)
    hand.push(byName.get(name) as T)
  }

  bags.set(key, bag)
  return hand
}

function shuffle<T>(items: readonly T[]): T[] {
  const out = [...items]
  for (let i = out.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[out[i], out[j]] = [out[j], out[i]]
  }
  return out
}

File: internal/subagent/starters.go
package subagent

// How a worker opens a conversation, kept in the worker's own folder.
//
// An empty chat shows a question and a few cards that fill the composer when
// clicked. Those used to be a list inside the desktop app, one set per room,
// which meant the one kind of worker the user can actually add — an agent is a
// folder (config.AgentHome) — was the one kind that could never have its own.
// Hiring a bookkeeper gave you four generic cards forever, because giving it
// four real ones meant editing the product (owner's call, 2026-08-10).
//
// So it is a file in the package, beside AGENT.md and skills/: STARTERS.md,
// with a translation beside it as STARTERS.<lang>.md.
//
//	# จะให้ทำเอกสารอะไรดี?
//
//	- ร่างรายงานหรือจดหมาย | ช่วยร่างเอกสารเรื่องนี้ให้หน่อย: | pencil
//	- ตรวจว่าเอกสารนี้ขาดอะไร | ช่วยตรวจเอกสารนี้ว่าขาดอะไรบ้าง: | search
//
// The format is markdown that happens to parse rather than a format with a
// spec: the heading is the question, each list item is a card, and the fields
// are split on `|` — title, the sentence that lands in the composer, and
// optionally the mark it wears. Someone reading the file in any editor sees a
// heading and a list, which is the point; a worker's opening line should not
// need a parser to be legible.
//
// What this package does NOT do is decide anything about the window. It reads
// a file and hands back what was in it. Where the cards are drawn, how many the
// grid fits, and what an empty answer falls back to are the window's business,
// and keeping that line is what lets a user's worker ship cards without the
// window having heard of it.

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/Mike0165115321/Aetox/internal/config"
)

// Starter is one card: what it says, what it puts in the composer, and the
// mark it wears. Icon is a name from the app's own icon set, same as the
// profile's `icon:` — a name rather than a picture, because the file is a .md
// somebody edits by hand.
type Starter struct {
	Title  string `json:"title"`
	Prompt string `json:"prompt"`
	Icon   string `json:"icon,omitempty"`
}

// StarterSet is a whole opening: the question, and the cards under it. Both are
// optional — a file with only a heading is a worker that asks its own question
// and leaves the cards to the window, which is a legitimate thing to want.
type StarterSet struct {
	Headline string    `json:"headline,omitempty"`
	Cards    []Starter `json:"cards"`
}

// maxStarters is the ceiling this package will return. Not a layout rule —
// the window owns its grid — but a bound on a file the app did not write, so a
// worker cannot hand back three hundred cards and make the window deal with it.
// Extra lines are dropped from the end, which is the order the author wrote
// them in and therefore the order they would have chosen to lose.
//
// It is a POOL ceiling, not a hand. It was 4 while the window drew every card
// a file held, which made those two numbers the same number and meant a worker
// could only gain a fifth card by giving up one of its four. The window now
// deals four out of whatever it is handed (starters.ts), so the file is free to
// carry more than fits on screen — which is the point of a pool: a worker gets
// deeper without the empty chat getting busier. 24 is six hands, far past what
// anyone will hand-write, and still small enough that a runaway file cannot
// turn into a payload the window has to think about.
const maxStarters = 24

// Starters reads one worker's opening in the requested language.
//
// Two homes and two languages, resolved in that order: **the home decides
// first**. A user who wrote their own STARTERS.md for a worker they took over
// gets their file in every language — falling back to the bundled English of a
// definition they replaced would be answering as an agent that no longer
// exists. Within the winning home, the requested language wins and the base
// file fills in for a language the author never wrote.
//
// Nothing here is an error. A worker with no file, an unreadable file, a file
// of prose with no list in it — all of them mean "no opening of my own", which
// the window already knows how to handle, and none of them is worth refusing to
// open a chat over.
func Starters(name, locale string) StarterSet {
	// "" is the base file. Asked for second, and only when the requested
	// language is not already it.
	languages := []string{locale}
	if config.AgentStartersName(locale) != config.AgentStartersFile {
		languages = append(languages, "")
	}
	for _, read := range []func(string) ([]byte, bool){userStarters(name), bundledStarters(name)} {
		for _, lang := range languages {
			if raw, ok := read(lang); ok {
				return parseStarters(string(raw))
			}
		}
	}
	return StarterSet{Cards: []Starter{}}
}

// SaveStarters writes one worker's opening for one language, in the same
// markdown Starters reads.
//
// The settings page's door. Editing the file by hand stays exactly as valid —
// that is what the format is for — and this is the same act through a form, so
// what it writes has to be a file a person would have been happy to write:
// a heading, a blank line, and a list. Nothing here is a marker the parser
// needs and a human does not.
//
// An empty set removes the user's file rather than writing an empty one. That
// is the honest inverse: with no file the worker falls back to the shipped
// opening, or to the four cards the window draws for any colleague — which is
// what "I do not want my own opening" means, and is not the same as "my opening
// is blank".
//
// Only the user's home is ever written. A bundled worker's opening is edited by
// writing your own beside it, the same shape as AGENT.md, so pressing Save on a
// shipped agent gives you a copy to change rather than a fight with the binary.
func SaveStarters(name, locale string, set StarterSet) error {
	path, err := config.AgentStartersPath(name, locale)
	if err != nil {
		return err
	}
	body, err := serializeStarters(set)
	if err != nil {
		return err
	}
	if body == "" {
		// Already absent is the state being asked for, so a missing file is
		// success rather than something to report.
		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
			return err
		}
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(body), 0o644)
}

// serializeStarters is parseStarters' inverse, and refuses what it cannot
// round-trip.
//
// `|` is the field separator, so one inside a title or a prompt would come back
// as a different card than the one that was saved — the icon field filled with
// half a sentence. Refused by name rather than escaped: an escape would put a
// backslash into a file whose whole point is that it reads as ordinary
// markdown, and refusing tells the user something they can act on.
func serializeStarters(set StarterSet) (string, error) {
	headline := strings.TrimSpace(set.Headline)
	var lines []string
	for _, c := range set.Cards {
		title := strings.TrimSpace(c.Title)
		// The space inviteCompletion adds after a trailing colon is put back on
		// read and must not be written: it is trailing whitespace, which every
		// editor strips and this parser trims anyway.
		prompt := strings.TrimSpace(c.Prompt)
		icon := strings.TrimSpace(c.Icon)
		if title == "" || prompt == "" {
			continue // a half-filled row is a row the user has not finished, not an error
		}
		if strings.ContainsAny(title+prompt+icon, "|\n") {
			return "", fmt.Errorf("การ์ด %q มีเครื่องหมาย | หรือขึ้นบรรทัดใหม่ ซึ่งใช้ในไฟล์นี้ไม่ได้", title)
		}
		if len(lines) >= maxStarters {
			break
		}
		if icon != "" {
			lines = append(lines, "- "+title+" | "+prompt+" | "+icon)
		} else {
			lines = append(lines, "- "+title+" | "+prompt)
		}
	}
	if headline == "" && len(lines) == 0 {
		return "", nil
	}
	var b strings.Builder
	if headline != "" {
		b.WriteString("# " + headline + "\n")
		if len(lines) > 0 {
			b.WriteString("\n")
		}
	}
	for _, line := range lines {
		b.WriteString(line + "\n")
	}
	return b.String(), nil
}

// userStarters reads from <DataRoot>/agents/<name>/. An invalid name simply
// never finds anything — AgentStartersPath is the guard, and it is the same one
// every other path into a home goes through.
func userStarters(name string) func(string) ([]byte, bool) {
	return func(locale string) ([]byte, bool) {
		path, err := config.AgentStartersPath(name, locale)
		if err != nil {
			return nil, false
		}
		raw, err := os.ReadFile(path)
		return raw, err == nil
	}
}

// bundledStarters reads from the shipped package of the same name. Embedded
// paths are always slash-separated, whatever the host OS does.
func bundledStarters(name string) func(string) ([]byte, bool) {
	return func(locale string) ([]byte, bool) {
		if !validName(name) {
			return nil, false
		}
		raw, err := bundledProfiles.ReadFile(bundledAgentDir + "/" + name + "/" + config.AgentStartersName(locale))
		return raw, err == nil
	}
}

// parseStarters reads the file. Every line is one of four things — a heading, a
// card, a comment, or nothing — and anything it cannot make a card out of is
// skipped rather than reported: this is a file a person hand-edits, and a typo
// on line three must cost line three and not the whole opening.
func parseStarters(raw string) StarterSet {
	set := StarterSet{Cards: []Starter{}}
	for _, line := range strings.Split(strings.ReplaceAll(raw, "\r\n", "\n"), "\n") {
		line = strings.TrimSpace(line)
		switch {
		case line == "", strings.HasPrefix(line, "<!--"):
			continue
		case strings.HasPrefix(line, "#"):
			// The first heading is the question. Later ones are the author
			// organising their own file and are not a second question.
			if set.Headline == "" {
				set.Headline = strings.TrimSpace(strings.TrimLeft(line, "#"))
			}
			continue
		}
		if len(set.Cards) >= maxStarters {
			continue
		}
		if card, ok := parseStarter(line); ok {
			set.Cards = append(set.Cards, card)
		}
	}
	return set
}

// parseStarter reads one line into a card, or says it is not one.
//
// A line with no `|` is prose — a note the author left themselves — and not a
// card with a missing half. Guessing that the whole line was both the title and
// the prompt would put a paragraph on a button.
func parseStarter(line string) (Starter, bool) {
	for _, bullet := range []string{"- ", "* ", "+ "} {
		if after, found := strings.CutPrefix(line, bullet); found {
			line = strings.TrimSpace(after)
			break
		}
	}
	parts := strings.SplitN(line, "|", 3)
	if len(parts) < 2 {
		return Starter{}, false
	}
	card := Starter{
		Title:  strings.TrimSpace(parts[0]),
		Prompt: strings.TrimSpace(parts[1]),
	}
	if len(parts) == 3 {
		card.Icon = strings.TrimSpace(parts[2])
	}
	if card.Title == "" || card.Prompt == "" {
		return Starter{}, false
	}
	card.Prompt = inviteCompletion(card.Prompt)
	return card, true
}

// inviteCompletion puts back the space after a trailing colon.
//
// A prompt ending in ":" is the deliberate half-sentence — the card fills the
// composer and the user types the rest — and the space between the colon and
// the cursor is part of it. It cannot be written in the file: it is trailing
// whitespace, which editors strip, git flags, and this parser trims. So the
// colon is the author's way of saying "this one is unfinished", and the space
// is put back here rather than being something they have to defend.
func inviteCompletion(prompt string) string {
	if strings.HasSuffix(prompt, ":") || strings.HasSuffix(prompt, "：") {
		return prompt + " "
	}
	return prompt
}
